exist-core/src/main/java/org/exist/xquery/functions/util/DocumentNameOrId.java

```diff
@@ -1,215 +1,206 @@
+                    if (arg.getType() == Type.STRING || arg.getType() == Type.ANY_URI) {
+                        return Sequence.of(getFromDocument(XmldbURI.create(arg.getStringValue()), DocumentNameOrId::getAbsoluteResourceId));
+                    } else if (arg instanceof NodeProxy) {
+                        return Sequence.of(getFromDocument(((NodeProxy) arg).getOwnerDocument().getURI(), DocumentNameOrId::getAbsoluteResourceId));
+                    } else {
+                        return Sequence.EMPTY_SEQUENCE;
+                    }
+
+                case FSN_GET_RESOURCE_BY_ABSOLUTE_ID:
+                    return getResourceByAbsoluteId((IntegerValue)arg);
+
+                default:
+                    throw new XPathException(this, "No function: " + getName() + "#" + getSignature().getArgumentCount());
+
+            }
+        } catch (final PermissionDeniedException e) {
+            throw new XPathException(this, arg.getStringValue() + ": permission denied to read resource");
         } catch (final LockException le) {
             throw new XPathException(this, "Unable to lock resource", le);
         } catch (final IOException ioe) {
             throw new XPathException(this, "Unable to read binary resource", ioe);
-        } catch (final URISyntaxException e) {
-            throw new XPathException(this, "Invalid resource uri: " + args[0].getStringValue(), e);
-        } catch (final PermissionDeniedException e) {
-            throw new XPathException(this, args[0].getStringValue() + ": permission denied to read resource");
         }
-        
-        return result;
     }
 
-    private Sequence documentNameOrId(final DocumentImpl doc) throws XPathException {
-        Objects.nonNull(doc);
-
-        final Sequence result;
-        final QName fnName = getSignature().getName();
-        if(fnName.equals(QN_DOCUMENT_NAME)) {
-            result = new StringValue(doc.getFileURI().toString());
-        } else if(fnName.equals(QN_DOCUMENT_ID)) {
-            result = new IntegerValue(doc.getDocId(), Type.INT);
-        } else if(fnName.equals(QN_ABSOLUTE_RESOURCE_ID)) {
-            BigInteger absoluteId = BigInteger.valueOf(doc.getCollection().getId());
-            absoluteId = absoluteId.shiftLeft(32);
-            absoluteId = absoluteId.or(BigInteger.valueOf(doc.getDocId()));
-            absoluteId = absoluteId.shiftLeft(1);
-            absoluteId = absoluteId.or(BigInteger.valueOf(doc.getResourceType() & 1));
-            result = new IntegerValue(absoluteId, Type.INTEGER);
-        } else {
-            result = Sequence.EMPTY_SEQUENCE;
+    private @Nullable <T> T getFromDocument(final XmldbURI docUri, final Function<DocumentImpl, T> docFn) throws PermissionDeniedException {
+        try (final LockedDocument lockedDoc = context.getBroker().getXMLResource(docUri, LockMode.READ_LOCK)) {
+            if (lockedDoc == null) {
+                return null;
+            } else {
```

Comment:
Is else necessary here? Since the if-statement will always return when lockedDoc is null.

Reply:
It's not necessary, but it's a coding style that I follow. When in more complex or nested code, always express the `else` path to make the program flow easier to follow.

Reply:
We are looking at a single statement. 🕶️